drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,651 @@
+
+	nrf_uarte_rx_buffer_set(uarte, &data->rx_data, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	if (interrupts_active) {
+		u8_t dummy_byte = 0;
+
+		irq_enable(NRFX_IRQ_NUMBER_GET(uarte));
+
+		/* Set ENDTX event by sending empty frame */
+		nrf_uarte_tx_buffer_set(uarte, &dummy_byte, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+
+		/* switch off transmitter to save an energy */
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	}
+#endif
+
+	return 0;
+}
+
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	static struct uarte_nrfx_data uarte_##idx##_data;		      \
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	      \
+		.max_chunk_len = (1 << UARTE##idx##_EASYDMA_MAXCNT_SIZE) - 1, \
+		.uarte.p_reg = (NRF_UARTE_Type *)NRF_UARTE##idx		      \
+	};								      \
+	static int uarte_##idx##_init(struct device *dev)		      \
+	{								      \
+		const struct uarte_init_config init_config = {		      \
+		.pseltxd = CONFIG_UART_##idx##_NRF_TX_PIN,		      \
+		.pselrxd = CONFIG_UART_##idx##_NRF_RX_PIN,		      \
+		.pselcts = CONFIG_UART_##idx##_NRF_CTS_PIN,		      \
+		.pselrts = CONFIG_UART_##idx##_NRF_RTS_PIN,		      \
+		.hwfc = CONFIG_UART_##idx##_NRF_FLOW_CONTROL,		      \
+		.parity = CONFIG_UART_##idx##_NRF_PARITY_BIT,		      \
+		.baudrate = CONFIG_UART_##idx##_BAUD_RATE		      \
+		};							      \
+		if (CONFIG_UART_##idx##_INTERRUPT_DRIVEN) {		      \
+			IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_UARTE##idx),      \
+				    CONFIG_UART_##idx##_IRQ_PRI,              \
+				    uarte_nrfx_isr,                           \
+				    DEVICE_GET(uart_nrfx_uarte##idx),         \
+				    0);                                       \
+		}							      \
+		return uarte_instance_init(dev,				      \
+				&init_config,				      \
+				CONFIG_UART_##idx##_INTERRUPT_DRIVEN);        \
+	}								      \
+	DEVICE_AND_API_INIT(uart_nrfx_uarte##idx,			      \
+			    CONFIG_UART_##idx##_NAME,                         \
+			    uarte_##idx##_init,                               \
+			    &uarte_##idx##_data,                              \
+			    &uarte_##idx##_config,                            \
+			    PRE_KERNEL_1,                                     \
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,               \
+			    &uart_nrfx_uarte_driver_api)
+
+#define UART_NRFX_UARTE_INVALID				(0xFFFFFFFFu)
+
+#ifdef CONFIG_UART_0_NRF_UARTE
+	DEVICE_DECLARE(uart_nrfx_uarte0);
+	#ifndef CONFIG_UART_0_INTERRUPT_DRIVEN
+		#define CONFIG_UART_0_INTERRUPT_DRIVEN	(0u)
+	#else
+		#undef CONFIG_UART_0_INTERRUPT_DRIVEN
+		#define CONFIG_UART_0_INTERRUPT_DRIVEN	(1u)
+	#endif /* UART_0_INTERRUPT_DRIVEN */
+	#ifndef CONFIG_UART_0_IRQ_PRI
+		#define CONFIG_UART_0_IRQ_PRI		UART_NRFX_UARTE_INVALID
+	#endif
+	#ifdef CONFIG_UART_0_NRF_FLOW_CONTROL
+		#undef CONFIG_UART_0_NRF_FLOW_CONTROL
+		#define CONFIG_UART_0_NRF_FLOW_CONTROL	NRF_UARTE_HWFC_ENABLED
```

Comment:
What about something like this?
```
#ifdef CONFIG_UART_0_NRF_FLOW_CONTROL
	#define UART_0_NRF_HWFC_CONFIG                \
		.pselcts = CONFIG_UART_0_NRF_CTS_PIN, \
		.pselrts = CONFIG_UART_0_NRF_RTS_PIN, \
		.hwfc = NRF_UARTE_HWFC_ENABLED,
#else
	#define UART_0_NRF_HWFC_CONFIG                  \
		.pselcts = NRF_UARTE_PSEL_DISCONNECTED, \
		.pselrts = NRF_UARTE_PSEL_DISCONNECTED, \
		.hwfc = NRF_UARTE_HWFC_DISABLED,
#endif /* CONFIG_UART_0_NRF_FLOW_CONTROL */
```
Or maybe even:
```
#define UART_NRF_HWFC_ENABLED(idx)                  \
	.pselcts = CONFIG_UART_##idx##_NRF_CTS_PIN, \
	.pselrts = CONFIG_UART_##idx##_NRF_RTS_PIN, \
	.hwfc = NRF_UARTE_HWFC_ENABLED,
#define UART_NRF_HWFC_DISABLED                  \
	.pselcts = NRF_UARTE_PSEL_DISCONNECTED, \
	.pselrts = NRF_UARTE_PSEL_DISCONNECTED, \
	.hwfc = NRF_UARTE_HWFC_DISABLED,
```
and then:
```
#ifdef CONFIG_UART_0_NRF_FLOW_CONTROL
	#define UART_0_NRF_HWFC_CONFIG  UART_NRF_HWFC_ENABLED(0)
#else
	...
```

Reply:
done